### PR TITLE
add gh release workflow

### PR DESCRIPTION
**Pull request by jkhelil**
# Changes
Adds github nightly workflow to build the project nighty and pushes a related image and nightly release.yaml for installation

# Submitter Checklist

- [*] Includes tests if functionality changed/was added
- [*] Includes docs if changes are user-facing
- [*] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)  
- [*] Release notes block has been filled in, or marked NONE

See [the contributor guide](https://github.com/shipwright-io/build/blob/main/CONTRIBUTING.md)
for details on coding conventions, github and prow interactions, and the code review process.

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by jkhelil** [comment]
@otaviof  can you have a look please

File: .github/workflows/nightly.yaml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    - uses: actions/setup-go@v3
+      with:
+        go-version: '1.19.x'
+        cache: true
+        check-latest: true
+
+    # Install tools
+    - uses: ko-build/setup-ko@v0.6
+      with:
+        version: v0.13.0
```
Comment (otaviof) [inline .github/workflows/nightly.yaml:32]: Please make use of the [common setup action](https://github.com/shipwright-io/triggers/blob/ab1363587736111e726e69394a246d0a69320820/.github/actions/setup/action.yaml), a usage example is the following:

https://github.com/shipwright-io/triggers/blob/ab1363587736111e726e69394a246d0a69320820/.github/workflows/test.yaml#L20

File: .github/workflows/nightly.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    - name: Get current date
+      id: date
+      run: echo "date=$(date +'%Y-%m-%d-%s')" >> $GITHUB_OUTPUT
+
+    - name: Install helm
+      run: |
+          curl -LO https://get.helm.sh/helm-v3.8.0-linux-amd64.tar.gz
+          tar -zxvf helm-v3.8.0-linux-amd64.tar.gz
+          mv linux-amd64/helm /usr/local/bin/helm
+          helm version
```
Comment (otaviof) [inline .github/workflows/nightly.yaml:46]: Please make use of the [azure/setup-helm](https://github.com/Azure/setup-helm) action.

File: Makefile [hunk 3]
```diff
@@ -97,7 +98,18 @@ deploy:
...
+
+# uses helm to create a package and ko for the container image
+.PHONY: release
+release: CGO_ENABLED=0
+release:
+	helm template \
+		--namespace=$(NAMESPACE) \
+		--set="image.name=ko://github.com/shipwright-io/triggers" \
+		shipwright-triggers \
+		$(CHART_DIR)| \
```
Comment (otaviof) [inline Makefile:111]: Let export `helm template` as a reusable macro, as in:

```Makefile
# uses helm to render the resource templates to the stdout
define render-template
    @helm template \
		--namespace=$(NAMESPACE) \
		--set="image.name=ko://github.com/shipwright-io/triggers" \
		shipwright-triggers \
		$(CHART_DIR)
endef
```
Having it defined, you can use on the `deploy` and `release` targets as the following snippet:

```Makefile
.PHONY: release
release: CGO_ENABLED=0
release:
    $(call render-template) | \
        ko resolve $(KO_OPTS) $(ARGS) \
            --platform=all \
            --recursive \
            --filename - >release.yaml
```

Also, make sure you're trying out the targets on your workstation, please. The `release` target contains a non-close quote for the `GOFLAGS` inline variable.

File: Makefile [hunk 4]
```diff
@@ -19,7 +19,8 @@ CGO_ENABLED ?= 0
 NAMESPACE ?= shipwright-build
 
 # ko base image repository and options
-KO_DOCKER_REPO ?= $(IMAGE_BASE)
+IMAGE_HOST ?= ghcr.io
+IMAGE_NAMESPACE ?= shipwright-io/triggers
```
Comment (otaviof) [inline Makefile:23]: Please leave the `KO_DOCKER_REPO` as a standalone variable, it get [exported as a regular environment variable using `.EXPORT_ALL_VARIABLES` target](https://github.com/shipwright-io/triggers/blob/3985967c8a670b2566cbc9f63c06fa32e199c642/Makefile#L48).

Thus, I suggest the following construction:

```Makefile
IMAGE_HOST ?= ghcr.io
IMAGE_NAMESPACE ?= shipwright-io/triggers
KO_DOCKER_REPO ?= $(IMAGE_HOST)/$(IMAGE_NAMESPACE)
```

The same approach takes place on `GOFLAGS`, no need to repeat yourself.

File: Makefile [hunk 5]
```diff
@@ -86,7 +87,7 @@ run: manifests
 .PHONY: container-build
 container-build: CGO_ENABLED=0
 container-build:
-	ko build --push=false $(KO_OPTS) $(ARGS) .
+	KO_DOCKER_REPO="$(IMAGE_HOST)/$(IMAGE_NAMESPACE)" GOFLAGS="${GOFLAGS} ko build --push=false $(KO_OPTS) $(ARGS) .
```
Comment (otaviof) [inline Makefile:90]: Same as the previous comment mentioning `.EXPORT_ALL_VARIABLES` behavior.

File: .github/workflows/nightly.yaml
Comment (otaviof) [inline .github/workflows/nightly.yaml:1]: Please use the same YAML format standards than the other workflow files.

File: .github/workflows/nightly.yaml [hunk 6]
```diff
@@ -0,0 +1,68 @@
+name: Nightly Release
```
Comment (otaviof) [inline .github/workflows/nightly.yaml:1]: Following the `test` workflow, please use `release-nigtly` instead.

**Review by otaviof — CHANGES_REQUESTED** [review]
Please consider the review comments. 🙏

Let's raise the community discussion of a nightly release. I think we should start releasing tags instead, and only later nightly changes.

/cc @adambkaplan @qu1queee @SaschaSchwarze0

File: .github/workflows/nightly.yaml (hunk 1 above)
Comment (jkhelil) [reply .github/workflows/nightly.yaml]: fixed

File: .github/workflows/nightly.yaml (hunk 2 above)
Comment (jkhelil) [reply .github/workflows/nightly.yaml]: fixed

File: Makefile (hunk 3 above)
Comment (jkhelil) [reply Makefile]: fixed

File: Makefile (hunk 4 above)
Comment (jkhelil) [reply Makefile:23]: fixed

File: .github/workflows/nightly.yaml
Comment (jkhelil) [reply .github/workflows/nightly.yaml:1]: fixed

File: Makefile (hunk 5 above)
Comment (jkhelil) [reply Makefile:100]: fixed

**Comment by jkhelil** [comment]
@otaviof I did some changes here and yes I agree, we need more a releasing workflow than a nightly for now
I add it similarly with what we have on build repo

File: Makefile (hunk 4 above)
Comment (otaviof) [reply Makefile:23]: Thanks! 👍

File: Makefile (hunk 5 above)
Comment (otaviof) [reply Makefile:100]: The variables `KO_DOCKER_REPO` and `GOFLAGS` are already exported globally (i.e.: `.EXPORT_ALL_VARIABLES` target). Please remove them from the local target.

File: Makefile [hunk 7]
```diff
@@ -86,19 +97,22 @@ run: manifests
...
-		shipwright-triggers \
-		$(CHART_DIR) | \
+	$(call render-template) | \
 			ko apply $(KO_OPTS) $(ARGS) --filename -
 
+# uses helm to create a package and ko for the container image
+.PHONY: release
+release: CGO_ENABLED=0
+release:
+	$(call render-template) | \
```
Comment (otaviof) [inline Makefile:113]: That's nice! 👍

**Review by otaviof — CHANGES_REQUESTED** [review]
A minor change needed, but overall looks good!

File: .github/draft_release_notes.sh [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+# it produces a 'Changes.md' file as its final output;
+# the file 'last-300-prs-with-release-note.txt' that is produces is intermediate data; it is not
+# pruned for now to assist development of the release notes process (we are still curating all this)
+
+if [ -z ${GITHUB_TOKEN+x} ]; then
+  echo "Error: GITHUB_TOKEN is not set"
+fi
+if [ -z ${PREVIOUS_TAG+x} ]; then
+  echo "Error: PREVIOUS_TAG is not set"
+fi
```
Comment (otaviof) [inline .github/draft_release_notes.sh:16]: Should it `exit 1` when the expected variables are not set?

File: .github/workflows/release.yaml [hunk 9]
```diff
@@ -0,0 +1,72 @@
+name: Release
```
Comment (otaviof) [inline .github/workflows/release.yaml:1]: nit: please [use lowercase](https://github.com/shipwright-io/triggers/blob/ab1363587736111e726e69394a246d0a69320820/.github/workflows/test.yaml#L2) here.

File: event.json
Comment (otaviof) [inline event.json:1]: Left over file? Please check if there are more files, like the `Changes.md` that should be added on `.gitignore`.

**Review by otaviof — CHANGES_REQUESTED** [review]
A few questions about handling errors, and a file left over.

File: .github/draft_release_notes.sh (hunk 8 above)
Comment (jkhelil) [reply .github/draft_release_notes.sh:16]: Not needed, the inputs are mandatory
```
inputs:
      release:
        description: 'Desired tag'
        required: true 
      tags:
        description: 'Previous tag'
        required: true
```

**Comment by jkhelil** [comment]
@otaviof I have fixed the few remaning errors, please fill free to review/approve

File: .github/workflows/release.yaml [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+      with:
+        version: v3.8.0
+
+    - name: Build Release Changelog
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        PREVIOUS_TAG: ${{ github.event.inputs.tags }}
+      run: |
+        # might not be necessary but make sure
+        chmod +x "${GITHUB_WORKSPACE}/.github/draft_release_notes.sh"
```
Comment (SaschaSchwarze0) [inline .github/workflows/release.yaml:40]: I agree with this comment. The file should be marked executable in Git and then this goes away.

**Review by SaschaSchwarze0 — CHANGES_REQUESTED** [review]

**Review by SaschaSchwarze0 — APPROVED** [review]
/approve
/lgtm

**Pull request merged**